x-pack/plugin/autoscaling/src/main/java/org/elasticsearch/xpack/autoscaling/capacity/AutoscalingCalculateCapacityService.java

```diff
@@ -30,14 +33,40 @@
 import java.util.stream.Collectors;
 import java.util.stream.StreamSupport;
 
-public class AutoscalingCalculateCapacityService {
-    private Map<String, AutoscalingDeciderService<? extends AutoscalingDeciderConfiguration>> deciderByName;
+public class AutoscalingCalculateCapacityService implements PolicyValidator {
+    private final Map<String, AutoscalingDeciderService> deciderByName;
 
-    public AutoscalingCalculateCapacityService(Set<AutoscalingDeciderService<? extends AutoscalingDeciderConfiguration>> deciders) {
+    public AutoscalingCalculateCapacityService(Set<AutoscalingDeciderService> deciders) {
         assert deciders.size() >= 1; // always have fixed
         this.deciderByName = deciders.stream().collect(Collectors.toMap(AutoscalingDeciderService::name, Function.identity()));
     }
 
+    public void validate(AutoscalingPolicy policy) {
+        policy.deciders().forEach(this::validate);
+    }
+
+    private void validate(final String deciderName, final Settings configuration) {
+        AutoscalingDeciderService deciderService = deciderByName.get(deciderName);
+        if (deciderService == null) {
+            throw new IllegalArgumentException("unknown decider [" + deciderName + "]");
+        }
+
+        Map<String, Setting<?>> deciderSettings = deciderService.deciderSettings()
+            .stream()
+            .collect(Collectors.toMap(s -> s.getKey(), Function.identity()));
+
+        configuration.keySet().forEach(key -> validateSetting(key, configuration, deciderSettings, deciderName));
+    }
+
+    private void validateSetting(String key, Settings configuration, Map<String, Setting<?>> deciderSettings, String decider) {
+        Setting<?> setting = deciderSettings.get(key);
+        if (setting == null) {
+            throw new IllegalArgumentException("unknown setting [" + key + "] for decider [" + decider + "]");
+        }
+
+        setting.get(configuration);
```

Comment:
Maybe a comment to indicate this method throws when there's an issue with the setting, and that's what we are testing for here? That is, make the purpose of this line clear?

Reply:
++, added in b9d87c5676afe7cf730559423767849443df61db